### PR TITLE
refactored filters to use icons

### PR DESCRIPTION
**Pull request by markgrahamdawson**
Closes https://github.com/cylc/cylc-ui/issues/1253

## The issue

The task filter v-select is used in both the tree and table views as well as for the gscan component. Previously this has been implemented separately in the two places.

There is also a workflow filter on on the gscan component which should be consistent in style and functionality with the task filter.

For the tree/table views the v-select used task icons.
image

For the gscan component the v-select used closable chips containing the names of the tasks/workflows
image

## The solution

I moved the functionality into a single v-select component (TaskFilterSelect.vue). 
This can be used for tasks and workflows - configured by props.

image

image

There is a bug with the way the vuetify v-select works in that it opens and closes the drop down whenever a value is selected - this was present before the refactor in this pr. I have opened a new issue to look into this further.
https://github.com/cylc/cylc-ui/issues/1535

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [ ] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [ ] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: src/components/cylc/TaskFilterSelect.vue [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+defineEmits(['update:modelValue'])
+
+const localValue = ref()
+localValue.value = computed(() => props.modelValue)
+
+const maxVisibleStates = 4
+
+function removeItem (key) {
+  localValue.value.value.states = localValue.value.value.states.filter(item => item !== key)
+}
```
Comment (markgrahamdawson) [inline src/components/cylc/TaskFilterSelect.vue:57]: Still to do: Should figure out what this value.value is all about :confused:

File: tests/e2e/specs/tree.cy.js [hunk 2]
```diff
@@ -218,7 +218,7 @@ describe('Tree view', () => {
           .contains(name)
```
Comment (markgrahamdawson) [inline tests/e2e/specs/tree.cy.js:218]: Im not sure if this really needs additional tests. All the old tests for GScan - tree and table are passing

File: src/components/cylc/TaskFilter.vue [hunk 3]
```diff
@@ -37,48 +37,20 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-          </div>
-          <span
-            v-if="index === $options.maxVisibleStates"
-            class="text-grey text-caption"
-          >
-            (+{{ localValue.states.length - $options.maxVisibleStates }})
-          </span>
-        </template>
-      </v-select>
+    <TaskFilterSelect :modelValue="localValue" :type="'task state'" :items="$options.allStates"></TaskFilterSelect>
```
Comment (MetRonnie) [inline src/components/cylc/TaskFilter.vue:40]: The conventional Vue style is to wrap if there is more than 1 attribute per tag: https://eslint.vuejs.org/rules/max-attributes-per-line.html (at some point we should probably turn on more `eslint-plugin-vue` rules).

Please could you go through and wrap these as I think the readability benefits
```suggestion
    <TaskFilterSelect
      :modelValue="localValue"
      type="task state"
      :items="$options.allStates"
    />
```

File: src/components/cylc/gscan/GScan.vue [hunk 4]
```diff
@@ -59,22 +59,11 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-                <v-select
-                  v-model="filters[title]"
-                  :items="$options.allStates[title]"
-                  :label="`Filter by ${title}`"
-                  density="default"
-                  chips
-                  closable-chips
-                  clearable
-                  multiple
-                  class="my-2"
```
Comment (MetRonnie) [inline src/components/cylc/gscan/GScan.vue:75]: The loss of this margin is causing display issues

image

File: src/components/cylc/TaskFilterSelect.vue [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  </v-chip>
+  <span
+      v-if="index === maxVisibleStates"
+      class="text-grey text-caption"
+    >
+      (+{{ localValue.states.length - maxVisibleStates }})
+  </span>
+</template>
+</v-select>
```
Comment (MetRonnie) [inline src/components/cylc/TaskFilterSelect.vue:34]: There is some missing indentation here

Comment (oliver-sanders) [reply src/components/cylc/TaskFilterSelect.vue:34]: Yet linting doesn't complain?

Comment (MetRonnie) [reply src/components/cylc/TaskFilterSelect.vue:34]: #1536 - it's included in `eslint-plugin-vue` but only if you turn on "strongly-recommended"

File: src/components/cylc/TaskFilterSelect.vue [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+<script setup>
+import { computed } from 'vue'
+import Task from '@/components/cylc/Task.vue'
+import Workflowicon from '@/components/cylc/gscan/WorkflowIcon.vue'
+
+const props = defineProps({
+  modelValue: Object, // { id, states }
+  items: Array,
+  type: Text
```
Comment (MetRonnie) [inline src/components/cylc/TaskFilterSelect.vue:45]: Getting
> [Vue warn]: Invalid prop: type check failed for prop "type". Expected Text, got String with value "task state"

```suggestion
  type: String
```

File: src/components/cylc/TaskFilter.vue [hunk 7]
```diff
@@ -54,7 +58,10 @@ export default {
   },
 
   props: {
-    modelValue: Object // { id, states }
+    modelValue: {
+      type: Object,
+      default: () => ({})
```
Comment (oliver-sanders) [inline src/components/cylc/TaskFilter.vue:63]: If this prop is required, you can set `required: true` rather than using a blank default.

**Review by MetRonnie — COMMENTED** [review]
I've got some suggestions I've spun out into https://github.com/markgrahamdawson/cylc-ui/pull/1

File: src/components/cylc/TaskFilterSelect.vue [hunk 8]
```diff
@@ -0,0 +1,95 @@
+<template>
+  <v-select
+    :label="`Filter by ${type}`"
+    :data-cy="`filter ${type}`"
+    :items="items"
+    clearable
+    multiple
+    :placeholder="`Filter by ${type}`"
```
Comment (MetRonnie) [inline src/components/cylc/TaskFilterSelect.vue:8]: In some cases we want to have the label and in others we want to have the placeholder. So if we don't specify these props they will be [fallthrough attributes](https://vuejs.org/guide/components/attrs.html) and you can specify them on the `<TaskFilerSelect>` tag instead

File: src/components/cylc/TaskFilter.vue [hunk 9]
```diff
@@ -37,52 +37,31 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-          <span
-            v-if="index === $options.maxVisibleStates"
-            class="text-grey text-caption"
-          >
-            (+{{ localValue.states.length - $options.maxVisibleStates }})
-          </span>
-        </template>
-      </v-select>
+        type="task state"
+        :items="$options.allStates"
```
Comment (MetRonnie) [inline src/components/cylc/TaskFilter.vue:43]: The text field before this uses placeholder not label, so for consistency 
```suggestion
        :items="$options.allStates"
        placeholder="Filter by task state"
```

File: src/components/cylc/gscan/GScan.vue [hunk 10]
```diff
@@ -26,97 +26,97 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+          <v-list>
+            <v-list-item
+              v-for="(items, title) in filters"
+              :key="title"
+            >
+              <TaskFilterSelect
+                v-model="filters[title]"
+                :type="title"
+                :items="$options.allStates[title]"
+                class="my-2"
```
Comment (MetRonnie) [inline src/components/cylc/gscan/GScan.vue:72]: To match how it looked previously
```suggestion
                :label="`Filter by ${title}`"
                density="default"
                class="my-2"
```

**Review by MetRonnie — APPROVED** [review]

**Review by oliver-sanders — APPROVED** [review]
:+1:

**Pull request merged**